File: app/src/main/java/com/drdisagree/iconify/xposed/mods/HookCheck.java
package com.drdisagree.iconify.xposed.mods;

import static android.content.Context.RECEIVER_EXPORTED;
import static com.drdisagree.iconify.common.Const.ACTION_HOOK_CHECK_REQUEST;
import static com.drdisagree.iconify.common.Const.ACTION_HOOK_CHECK_RESULT;
import static com.drdisagree.iconify.common.Const.SYSTEMUI_PACKAGE;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;

import com.drdisagree.iconify.xposed.ModPack;

import java.util.Objects;

import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class HookCheck extends ModPack {

    private static final String TAG = "Iconify - " + HookCheck.class.getSimpleName() + ": ";
    IntentFilter intentFilter = new IntentFilter();
    boolean broadcastRegistered = false;

    public HookCheck(Context context) {
        super(context);
    }

    @Override
    public void updatePrefs(String... Key) {
    }

    private void returnBroadcastResult() {
        new Thread(() -> mContext.sendBroadcast(new Intent()
                .setAction(ACTION_HOOK_CHECK_RESULT)
                .addFlags(Intent.FLAG_RECEIVER_FOREGROUND))).start();
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        if (!broadcastRegistered && lpparam.packageName.equals(SYSTEMUI_PACKAGE)) {
            broadcastRegistered = true;
            intentFilter.addAction(ACTION_HOOK_CHECK_REQUEST);

            BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    if (Objects.equals(intent.getAction(), ACTION_HOOK_CHECK_REQUEST) && lpparam.packageName.equals(SYSTEMUI_PACKAGE)) {
                        returnBroadcastResult();
                    }
                }
            };

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                mContext.registerReceiver(broadcastReceiver, intentFilter, RECEIVER_EXPORTED);
            } else {
                mContext.registerReceiver(broadcastReceiver, intentFilter);
            }
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/xposed/utils/HookCheck.kt
package com.drdisagree.iconify.xposed.utils

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import com.drdisagree.iconify.BuildConfig
import com.drdisagree.iconify.data.common.Const.ACTION_HOOK_CHECK_REQUEST
import com.drdisagree.iconify.data.common.Const.ACTION_HOOK_CHECK_RESULT
import com.drdisagree.iconify.data.common.Const.FRAMEWORK_PACKAGE
import com.drdisagree.iconify.data.common.Const.SYSTEMUI_PACKAGE
import com.drdisagree.iconify.xposed.HookEntry
import com.drdisagree.iconify.xposed.ModPack
import io.github.libxposed.api.XposedModuleInterface.PackageReadyParam

class HookCheck(context: Context) : ModPack(context) {

    private var intentFilter = IntentFilter()
    private var broadcastRegistered = false

    override fun onPreferenceUpdated(vararg key: String) {}

    private fun returnBroadcastResult() {
        Thread {
            mContext.sendBroadcast(
                Intent()
                    .putExtra(
                        "packageName", if (HookEntry.isSystemServer) FRAMEWORK_PACKAGE
                        else SYSTEMUI_PACKAGE
                    )
                    .setPackage(BuildConfig.APPLICATION_ID)
                    .setAction(ACTION_HOOK_CHECK_RESULT)
                    .addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
            )
        }.start()
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    override fun onPackageLoaded(packageReadyParam: PackageReadyParam) {
        if (!broadcastRegistered) {
            broadcastRegistered = true

            intentFilter.addAction(ACTION_HOOK_CHECK_REQUEST)

            val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    if (intent.action == ACTION_HOOK_CHECK_REQUEST) {
                        returnBroadcastResult()
                    }
                }
            }

            mContext.registerReceiver(
                broadcastReceiver,
                intentFilter,
                Context.RECEIVER_EXPORTED
            )
        }
    }
}
